src/librustc_interface/util.rs

```diff
@@ -289,20 +289,34 @@ pub fn get_codegen_backend(sess: &Session) -> Box<dyn CodegenBackend> {
     backend
 }
 
-pub fn get_codegen_sysroot(backend_name: &str) -> fn() -> Box<dyn CodegenBackend> {
-    // For now we only allow this function to be called once as it'll dlopen a
-    // few things, which seems to work best if we only do that once. In
-    // general this assertion never trips due to the once guard in `get_codegen_backend`,
-    // but there's a few manual calls to this function in this file we protect
-    // against.
-    static LOADED: AtomicBool = AtomicBool::new(false);
-    assert!(!LOADED.fetch_or(true, Ordering::SeqCst),
-            "cannot load the default codegen backend twice");
+// This is used for rustdoc, but it uses similar machinery to codegen backend
+// loading, so we leave the code here. It is potentially useful for other tools
+// that want to invoke the rustc binary while linking to rustc as well.
+pub fn rustc_path() -> Option<&Path> {
+    lazy_static::lazy_static! {
+        static ref RUSTC_PATH: PathBuf = get_rustc_path_inner();
+    }
+
+    RUSTC_PATH.as_ref().map(|v| v)
+}
+
+fn get_rustc_path_inner() -> PathBuf {
+    sysroot_candidates().iter()
+        .filter_map(|sysroot| {
+            let candidate = sysroot.join("bin").join("rustc");
```

Comment:
This won't always be `"bin"`: https://github.com/rust-lang/rust/blob/122fefc63e4c2bab7f76bb24a7ab056b7bc7b37e/config.toml.example#L236-L237

Also `"rustc"` needs to be `"rustc.exe"` on Windows.

Reply:
Hm, I think the latter is certainly not true -- this was working previously with just rustc on all platforms, likely because of https://github.com/rust-lang/rust/issues/37519.

I can explicitly add `.exe` on windows and pull in the bindir specified in config.toml though

Reply:
It's the `.exists()` call below that won't work without the `.exe`.